core/src/muxing.rs

```diff
@@ -93,5 +104,132 @@ pub trait StreamMuxer {
     /// >           that the remote is properly informed of the shutdown. However, apart from
     /// >           properly informing the remote, there is no difference between this and
     /// >           immediately dropping the muxer.
-    fn poll_close(&self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>>;
+    fn poll_close(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>>;
+}
+
+/// Extension trait for [`StreamMuxer`].
+pub trait StreamMuxerExt: StreamMuxer + Sized {
+    fn poll_inbound_unpin(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<Self::Substream, Self::Error>>
+    where
+        Self: Unpin;
+
+    fn poll_outbound_unpin(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<Self::Substream, Self::Error>>
+    where
+        Self: Unpin;
+
+    fn poll_address_change_unpin(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<Multiaddr, Self::Error>>
+    where
+        Self: Unpin;
+
+    fn poll_close_unpin(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>>
+    where
+        Self: Unpin;
+
+    fn next_inbound(&mut self) -> NextInbound<'_, Self>;
+
+    fn next_outbound(&mut self) -> NextOutbound<'_, Self>;
+
+    fn close(self) -> Close<Self>;
+}
+
+impl<S> StreamMuxerExt for S
+where
+    S: StreamMuxer,
+{
+    fn poll_inbound_unpin(
```

Comment:
Why not provide this as a default implementation on the trait definition itself?

Reply:
I guess we could because we own the trait?

Default impls can be overridden and these are not really meant too. I think that is the only argument for a separate trait.

Reply:
No strong opinion here. One datapoint is that `StreamExt` implements the methods directly:

https://docs.rs/futures/latest/futures/prelude/stream/trait.StreamExt.html

Reply:
Oh, now I understand what you mean! I thought we are discussing having the extension trait vs putting the functions as defaults on `StreamMuxer`.

With the extension trait, I think I prefer default functions too!

Reply:
Great. To make sure we are on the same page, with the above in mind, this `trait` `impl` would go away @thomaseizinger?

Reply:
The `impl` block would be there but it will be a one-liner as all the implementations are in the trait definition.

One alternative idea would be to remove `StreamMuxerExt` entirely and add the default functions there directly.

I think I mildly prefer the extension trait though because:

- It makes it impossible for implementations to overwrite them, thus guaranteeing that they are the same for all.
- It makes `StreamMuxer` a more coherent trait and we could draw a dependency boundary between `StreamMuxer` and its extension trait if we wanted to.